src/ui/hook/usePrevious.ts

```diff
@@ -0,0 +1,9 @@
+import { useRef, useEffect } from 'react'
+
+export const usePrevious = <T>(value: T): T | undefined => {
+  const ref = useRef<T>()
+  useEffect(() => {
+    ref.current = value
+  })
+  return ref.current
```

Comment:
IMHO, it's better to update `ref` only when `value` changes, not each time the hook is invoked.
```suggestion
  useEffect(() => {
    ref.current = value
  }, [value])
  return ref.current
```